Add Documentation to the mfe_deployer role

morenol opened the pull request:
Configuration Pull Request
---

Make sure that the following steps are done before merging:

  - [ ] A DevOps team member has approved the PR if it is code shared across multiple services and you don't own all of the services.
  - [ ] Are you adding any new default values that need to be overridden when this change goes live? If so:
    - [ ] Update the appropriate internal repo (be sure to update for all our environments)
    - [ ] If you are updating a secure value rather than an internal one, file a DEVOPS ticket with details.
    - [ ] Add an entry to the CHANGELOG.
  - [ ] If you are making a complicated change, have you performed the proper testing specified on the [Ops Ansible Testing Checklist](https://openedx.atlassian.net/wiki/display/EdxOps/Ops+Ansible+Testing+Checklist)?  Adding a new variable does not require the full list (although testing on a sandbox is a great idea to ensure it links with your downstream code changes).
  - [ ] Think about how this change will affect Open edX operators.  Have you updated the wiki page for the next Open edX release?

edx-webhook commented:
Thanks for the pull request, @morenol! I've created [OSPR-5251](https://openedx.atlassian.net/browse/OSPR-5251) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

natabene commented:
@morenol Thank you for your contribution. Please let me know once it is ready for our review.

pomegranited commented:
@morenol  I'd like to review this as core committer, when you're ready? CC @natabene

giovannicimolin commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,66 @@
+
+For a generic overview of the specific steps that are required for the deployment, please see the [Developer Documentation](https://edx.readthedocs.io/projects/edx-developer-docs/en/latest/developers_guide/micro_frontends_in_open_edx.html).
+
+In order to deploy MFEs with ansible, there were created by the community two roles `mfe` and `mfe_deployer`:
```
```suggestion
Two ansible roles were created by the community to deploy MFEs using `edx/configuration`:
```

giovannicimolin commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,66 @@
+Configuration & Deployment
+--------------------------
+
+When running this role, you'' need to set:
```
```suggestion
When running this role, you'll need to set the following variables:
```

giovannicimolin commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,66 @@
+
+In order to deploy MFEs with ansible, there were created by the community two roles `mfe` and `mfe_deployer`:
+
+- `mfe`: The `mfe` role is the base role for the deployments, in that role a specific mfe is cloned, build and then its files are served using nginx. This role internally, follows the steps for building and deployiment of MFEs described in the developer documentation, with the exception that it does not uses tubular. Ansible users should not use directly this role, and instead use the `mfe_deploy ...
```
```suggestion
- `mfe`: The `mfe` role is the base role for the deployments where a single mfe is cloned, built and set up to be served using nginx from the appserver. This role internally, follows the steps for building and deployment of MFEs described in the developer documentation, with the exception that it does not uses tubular. Ansible users should not use directly this role, and instead use the `mfe_deployer` role.
```

giovannicimolin commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,66 @@
+
+When running this role, you'' need to set:
+
+- `MFES`: list that contains each one of the MFEs that will be deployed. Each MFE should contain a `name` and a `repo`. Optionally, it can be defined for each MFE:
```
```suggestion
- `MFES`: list that contains each one of the MFEs that will be deployed. Each MFE should contain a `name` and a `repo`. Optionally, the extra parameters below can be defined for each MFE:
```

giovannicimolin commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,66 @@
+        - `standalone_nginx`: To indicate if the MFE will be deployed in a separated nginx file or if it will be in a shared nginx file with the other MFEs, by default it takes the value of `MFE_DEPLOY_STANDALONE_NGINX`.
+
+
+By default, MFES is defined as:
```
```suggestion
By default, MFES variable is defined as follows:
```

giovannicimolin commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,66 @@
+
+So it will deploy three MFEs each one in a different subdirectory. Since the default value of `MFE_DEPLOY_STANDALONE_NGINX` is false. The three MFE will be deployed in the same nginx file.
+
+As it was described previously, the compilation of the MFEs is done in the `mfe` role. Some configurations cannot be override as of today from the mfe_deployer, you can see the list of all the default environment configuration in the defaults of the mfe role. For instance, it can be defined `MFE_MARKETING_SITE_BASE_URL`, `MFE_ENTERPRISE_MARKETING_UTM_SOURCE`, `MFE_ENTERPRISE_MARKETING_UTM_CAMPAIG ...
```
```suggestion
As described previously, the compilation of the MFEs is done in the `mfe` role, so some configurations cannot be overridden from `mfe_deployer`. You can see the list of all the default environment configuration in the defaults of the `mfe` role. The following variables can be overridden for all MFEs, but not individually: `MFE_MARKETING_SITE_BASE_URL`, `MFE_ENTERPRISE_MARKETING_UTM_SOURCE`, `MFE_ENTERPRISE_MARKETING_UTM_CAMPAIGN`, and `MFE_ENTERPRISE_MARKETING_FOOTER_UTM_MEDIUM` for configuration related to the marketing site and `MFE_NEW_RELIC_APP_ID` and `MFE_NEW_RELIC_LICENSE_KEY` in order to configure the newrelic integration.
```

giovannicimolin replied:
@morenol Can you add a link to the MFE default variable here?

giovannicimolin commented on playbooks/roles/mfe_deployer/README.rst:
````diff
@@ -0,0 +1,66 @@
+    public_path: "/account/
+```
+
+So it will deploy three MFEs each one in a different subdirectory. Since the default value of `MFE_DEPLOY_STANDALONE_NGINX` is false. The three MFE will be deployed in the same nginx file.
````
```suggestion
This configuration will deploy three MFEs, each in a different subdirectory. And since the default value of `MFE_DEPLOY_STANDALONE_NGINX` is false,  the three MFEs will be deployed in the same nginx file.
```

giovannicimolin replied:
@morenol I'm not sure what you mean by `deployed in the same Nginx file` here can you elaborate a bit more?

Also, the default value of `MFE_DEPLOY_STANDALONE_NGINX` is true, see https://github.com/edx/configuration/blob/b828714637f55fc2c11faec8e59b287c7fc730fd/playbooks/roles/mfe/defaults/main.yml#L136.

morenol replied:
You are right, I have to add a better description, I am trying to say that the nginx configuration for all  of the MFEs is placed in the same in 

/etc/nginx/sites_enabled/

I agree that I have to explain that in a more descriptive way. On the other hand it is true in that role, but in the mfe_deployer (which is the role that should be used by the edx/configuration  users) it is false.

giovannicimolin commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,66 @@
+
+
+Configuration & Deployment
+--------------------------
```
I think it's worth mentioning here that MFEs can be deployed in one of two ways:

1. Using the same subdomain, but served from different directories.
2. Using separate subdomains.

And mention the variables that control which deployment type is configured.
We can also mention what the default values deploy, and a example of variables to deploy on separate subdomains vs on the same domain.

giovannicimolin commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,66 @@
+LMS Configuration
+-----------------
+
+The deployment of the MFEs by themselves will not be enought to have them working properly. Most of them require communication with the LMS, so it is necessary to configure the LMS to accept communication from the MFEs.
```
```suggestion
The deployment of the MFEs to the appservers will not be enough to have them working properly. Most of them require communication with the LMS, so it is necessary to configure the LMS to accept communication from the MFEs.
```

giovannicimolin commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,66 @@
+
+The deployment of the MFEs by themselves will not be enought to have them working properly. Most of them require communication with the LMS, so it is necessary to configure the LMS to accept communication from the MFEs.
+
+The principal configurations that are needed in ansible are: `EDXAPP_CORS_ORIGIN_WHITELIST, EDXAPP_CSRF_TRUSTED_ORIGINS, EDXAPP_LOGIN_REDIRECT_WHITELIST`. They should contain the domain of the MFEs in order to configure the LMS to accepts requests from the MFE.
```
```suggestion
The principal configurations that are needed in ansible are: `EDXAPP_CORS_ORIGIN_WHITELIST, EDXAPP_CSRF_TRUSTED_ORIGINS, EDXAPP_LOGIN_REDIRECT_WHITELIST`. 
They should contain the domain of the MFEs so that the LMS accepts their requests.
```

giovannicimolin commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,66 @@
+
+It is also necessary to enable the features ENABLE_CORS_HEADERS and ENABLE_CROSS_DOMAIN_CSRF_COOKIE. They can be enabled in Koa with `EDXAPP_ENABLE_CORS_HEADERS` `EDXAPP_ENABLE_CROSS_DOMAIN_CSRF_COOKIE`.
+
+It is also necessary to have configured JWT properly in the LMS.
```
By this you mean that the JWK Key pairs are configured correctly? Is there any specific configuration?

morenol replied:
Yes, I think that it is not needed a specific configuration for them, only the configurations that we can generate through this command https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/oauth_dispatch/management/commands/generate_jwt_signing_key.py/

Other than that, I need to check better which are the required configurations for the CORS configurations.

giovannicimolin reviewed (changes requested):
@morenol Thanks a lot for pushing this work forward and documenting the MFE deployment.
This will help everyone that wants to deploy MFE's in the community. Great work!

I've made lots of comments/suggestions, feel free to reply/accept/reject them as you see fit.
I still have some free time for this in the next week, so ping me if you need to discuss any comments (or the general approach to documenting this feature).

giovannicimolin commented:
@morenol Ping either me or @pomegranited when you need another review round. Nice work so far! :grin:

morenol commented:
Hi @pomegranited @giovannicimolin, could you check this?

blarghmatey commented:
Looking at the two roles, it seems that `MFE_BASE` should also be overridden for the case where you are deploying using a shared Nginx configuration. I'm also interested in understanding whether this is compatible with running on the same instance and domain as edx-platform or if the expectation is that they will be separated by both domain and host?

It seems like in order for colocation with edx-platform the `MFE_BASE` would need to be set to the same value as `COMMON_LMS_BASE_URL` and that the configuration would then become the responsibility of the edx-platform. Otherwise the `server_name` and port would conflict between the two Nginx processes.

pomegranited reviewed (changes requested):
Thank you for preparing these docs @morenol !

I've noted some formatting issues with links and code blocks which you can see if you view the [rendered README files](https://github.com/edx/configuration/blob/56be97cfec1cffbda5bcf00ef4bf375f191b4084/playbooks/roles/mfe_deployer/README.rst).

But once those are addressed I think this is good to merge.

morenol commented:
> Looking at the two roles, it seems that `MFE_BASE` should also be overridden for the case where you are deploying using a shared Nginx configuration. 

Yes, you are right. I am going to add that to the deploying in subdirectories part.

> I'm also interested in understanding whether this is compatible with running on the same instance and domain as edx-platform or if the expectation is that they will be separated by both domain and host?
It seems like in order for colocation with edx-platform the `MFE_BASE` would need to be set to the same value as `COMMON_LMS_BASE_URL` and that the configuration would then become the responsibility of the edx-platform. Otherwise the `server_name` and port would conflict between the two Nginx processes.

I think that it is totally possible to run them in the same domain than the LMS, in fact I did that before.  As you said with this role it will only be possible to do that if they are separated in different hosts due to conflicts in the nginx configuration. We can use another service that routes the requests to the right server using the path of the requests.

pomegranited commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,102 @@
+
+By default, the mfe_deployer role is configured to deploy the microfrontends using subdirectories, the most important configurations to enable that are **MFES** and **MFE_DEPLOY_STANDALONE_NGINX**.  The *MFES* variable by default is defined as follows:
+
+    MFES:
```
I don't think it's a good idea to repeat the default configuration here. Instead, better to show complete example config which deploys the default MFEs to subdirectories of a single subdomain and hooks them up to the LMS, e.g. 

```yaml
MFE_BASE: "mfe.{{ EDXAPP_LMS_BASE }}"

EDXAPP_SESSION_COOKIE_DOMAIN: ".{{ EDXAPP_LMS_BASE }}"
EDXAPP_CSRF_COOKIE_SECURE: true
EDXAPP_SESSION_COOKIE_SECURE: true
EDXAPP_CROSS_DOMAIN_CSRF_COOKIE_DOMAIN: ".{{ EDXAPP_LMS_BASE }}"
EDXAPP_CROSS_DOMAIN_CSRF_COOKIE_NAME: "cross-domain-cookie-mfe"

# Whitelisting
EDXAPP_CORS_ORIGIN_WHITELIST:
- "{{ EDXAPP_CMS_BASE }}"
- "{{ MFE_BASE }}"

EDXAPP_CSRF_TRUSTED_ORIGINS:
- "{{ MFE_BASE }}"

EDXAPP_LOGIN_REDIRECT_WHITELIST:
- "{{ EDXAPP_CMS_BASE }}"
- "{{ MFE_BASE }}"

# MFE Links
EDXAPP_LMS_WRITABLE_GRADEBOOK_URL: https://{{ MFE_BASE}}/gradebook
EDXAPP_PROFILE_MICROFRONTEND_URL: https://{{ MFE_BASE}}/profile/u/
EDXAPP_ACCOUNT_MICROFRONTEND_URL: https://{{ MFE_BASE}}/account
```

pomegranited commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,102 @@
+The deployment of the MFEs to the appservers will not be enough to have them working properly. Most of them require communication with the LMS, so it is necessary to configure the LMS to accept communication from the MFEs.
+
+The principal configurations that are needed in ansible are: *EDXAPP_CORS_ORIGIN_WHITELIST*, *EDXAPP_CSRF_TRUSTED_ORIGINS*, *EDXAPP_LOGIN_REDIRECT_WHITELIST*. 
+They should contain the domain of the MFEs so that the LMS accepts their requests.
```
Worth mentioning the MFE URL variables that you also need to set so the LMS uses the default `MFES`: 
`EDXAPP_LMS_WRITABLE_GRADEBOOK_URL`, `EDXAPP_PROFILE_MICROFRONTEND_URL`,
`EDXAPP_ACCOUNT_MICROFRONTEND_URL`.

pomegranited commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,102 @@
+_______________________________
+
+
+By default, the mfe_deployer role is configured to deploy the microfrontends using subdirectories, the most important configurations to enable that are **MFES** and **MFE_DEPLOY_STANDALONE_NGINX**.  The *MFES* variable by default is defined as follows:
```
The [wording below](https://github.com/edx/configuration/pull/6171/files#diff-0539a91cf83ea97c0637aecf583609aae0a484c76158abc8809b4909d424df1dR56) is confusing and duplicates some information contained here.  Could these be simplified to something like this:

```suggestion
By default `MFE_DEPLOY_STANDALONE_NGINX` is false, which means that all the microfrontends defined in the `MFES` configuration will be deployed to different `public_path`s under the same domain (specifically, `MFE_DEPLOY_COMMON_HOSTNAME`).
```

pomegranited commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,102 @@
+        public_path: "/"
+      - name: account
+        repo: frontend-app-account
+        public_path: "/"
```
This code block and the ones above are still in MD format, see [rendered file](https://github.com/edx/configuration/blob/d78deab850f712bdf3a8dd720dd0b8a87e2062a0/playbooks/roles/mfe_deployer/README.rst#deployment-using-subdirectories).

pomegranited commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,102 @@
+        repo: frontend-app-account
+        public_path: "/"
+
+    MFE_DEPLOY_STANDALONE_NGINX: true
```
I think it's clearer to include the whitelisting and CSRF variables required for this example here, instead of just discussing them below.

```yaml
EDXAPP_SESSION_COOKIE_DOMAIN: ".{{ EDXAPP_LMS_BASE }}"
EDXAPP_CSRF_COOKIE_SECURE: true
EDXAPP_SESSION_COOKIE_SECURE: true
EDXAPP_CROSS_DOMAIN_CSRF_COOKIE_DOMAIN: ".{{ EDXAPP_LMS_BASE }}"
EDXAPP_CROSS_DOMAIN_CSRF_COOKIE_NAME: "cross-domain-cookie-mfe"

# Whitelisting
EDXAPP_LOGIN_REDIRECT_WHITELIST:
- "{{ EDXAPP_CMS_BASE }}"
- "profile.{{ EDXAPP_LMS_BASE }}"
- "gradebook.{{ EDXAPP_LMS_BASE }}"
- "account.{{ EDXAPP_LMS_BASE }}"

EDXAPP_CORS_ORIGIN_WHITELIST:
- "{{ EDXAPP_CMS_BASE }}"
- "profile.{{ EDXAPP_LMS_BASE }}"
- "gradebook.{{ EDXAPP_LMS_BASE }}"
- "account.{{ EDXAPP_LMS_BASE }}"

EDXAPP_CSRF_TRUSTED_ORIGINS:
- "{{ EDXAPP_CMS_BASE }}"
- "profile.{{ EDXAPP_LMS_BASE }}"
- "gradebook.{{ EDXAPP_LMS_BASE }}"
- "account.{{ EDXAPP_LMS_BASE }}"

# MFE Links
EDXAPP_LMS_WRITABLE_GRADEBOOK_URL: "https://gradebook.{{ EDXAPP_LMS_BASE}}/"
EDXAPP_PROFILE_MICROFRONTEND_URL: "https://profile.{{ EDXAPP_LMS_BASE}}/"
EDXAPP_ACCOUNT_MICROFRONTEND_URL: "https://account.{{ EDXAPP_LMS_BASE}}/"
```

pomegranited reviewed (changes requested):
There's still some formatting issues with [the rendered RST](https://github.com/edx/configuration/blob/d78deab850f712bdf3a8dd720dd0b8a87e2062a0/playbooks/roles/mfe_deployer/README.rst#deployment-using-subdirectories).

But thank you for adding the subsections for subdomain and subdirectory deployments! I left some comments requesting that those code blocks be complete example configurations, so that people can see at a glance what needs to be set to do the integration with the LMS.

pomegranited commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,148 @@
+    - "{{ MFE_BASE }}"
+
+  # MFE Links
+  EDXAPP_LMS_WRITABLE_GRADEBOOK_URL: 'https://{{ MFE_BASE}}/gradebook'
```
FYI unfortunately, there's a bug in edx-platform that prevents the gradebook link from working when deployed with subdirectories.

[The code](https://github.com/edx/edx-platform/blob/151bd136667d27bf85256e382ed5b5ea0144e00f/lms/djangoapps/instructor/views/instructor_dashboard.py#L574) uses [`urljoin`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urljoin), which uses only the hostname from `WRITABLE_GRADEBOOK_URL`, and removes the `/gradebook` path.

So the above example produces e.g., `https://mfe.my.lms.com/course-v1:edX+DemoX+Demo_Course` instead of `https://mfe.my.lms.com/gradebook/course-v1:edX+DemoX+Demo_Course`.

Not something that needs to be fixed here obviously, but noting it as something we'll need to fix to use subdirectories with this MFE.

pomegranited commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,148 @@
+  EDXAPP_LOGIN_REDIRECT_WHITELIST:
+    - "{{ EDXAPP_CMS_BASE }}"
+    - "{{ MFE_BASE }}"
+
```
Only thing missing for `CORS` is:

```yaml
EDXAPP_FEATURES:
  ENABLE_CORS_HEADERS: true
```

morenol replied:
I think that unless you are overriding the default EDXAPP_FEATURES configuration, it should be being working with `EDXAPP_ENABLE_CORS_HEADERS`

https://github.com/edx/configuration/blob/a2c4e00ec157b762ea5786111db22bf34e005742/playbooks/roles/edxapp/defaults/main.yml#L327

pomegranited replied:
Cool, yep, that's better!

pomegranited reviewed (changes requested):
@morenol Thank you so much for adding the "Comprehensive Example config for subdirectories", that's useful! I've validated it using [this sandbox](https://second-mfe-test.opencraft.hosting), and found only one thing missing, the `ENABLE_CORS_HEADERS` feature as noted below.

I also validated the subdomain config [described here](https://github.com/edx/configuration/pull/6171/files#r547091288) using [another sandbox](https://mfe-test.sandbox.opencraft.hosting). So if you can also add the comprehensive example config for subdomains too, then this is ready to approve and merge.

pomegranited commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,157 @@
+Deployment using subdirectories
+_______________________________
+
+By default ``MFE_DEPLOY_STANDALONE_NGINX`` is false, which means that all the microfrontends defined in the ``MFES`` configuration will be deployed to different ``public_path``'s under the same domain (specifically, ``MFE_DEPLOY_COMMON_HOSTNAME``). On the other hand, ``MFE_DEPLOY_STANDALONE_NGINX`` is false by default in the mfe_deployer role which means that all the microfrontends defined in the ...
```
nit: the same thing is said twice here, so you can remove the second sentence, starting with "On the other hand...".

pomegranited commented on playbooks/roles/mfe_deployer/README.rst:
```diff
@@ -0,0 +1,157 @@
+
+It is also necessary to have configured JWT properly in the LMS. You can use the generate_jwt_signing_key command to generate the signing key. See `decision record about asymmetric JWT`_ for more details.
+
+For each MFE, it might be certain configurations that also need to be changed according to the URLs of the MFE, for instance, for the gradebook, progile and account MFE we need to set ``EDXAPP_LMS_WRITABLE_GRADEBOOK_URL``, ``EDXAPP_PROFILE_MICROFRONTEND_URL`` and ``EDXAPP_ACCOUNT_MICROFRONTEND_URL`` with their respective URLs.
```
nit: "progile" is misspelling of "profile"

pomegranited reviewed (approved):
:+1: Thank you for your patience and persistence here @morenol ! Your docs look great, and make your excellent MFE deployment role contribution accessible and usable by the community.

I've found two more tiny nits noted below, but if you could make those changes and re-squash your commits, I'll merge.

- [x] I tested this by deploying sandboxes with the default MFEs using both mechanisms: [MFE subdomains](https://mfe-test.sandbox.opencraft.hosting) and [MFE subdirectories](https://second-mfe-test.opencraft.hosting/).
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike>
- [x] Includes documentation
- [x] <strike>I made sure any change in configuration variables is reflected in the corresponding
    client's `configuration-secure` repository.</strike> N/A

nedbat commented:
I am very impressed by the work you all are doing on this! Docs are critical but easy to skimp on.

@regisb does this address some of the issues you had in Tutor?

regisb commented:
> @regisb does this address some of the issues you had in Tutor?

Not really; I am more interested in learning about the MFE settings which are configured via environment variables. The default environment is stored in the [MFE_ENVIRONMENT_DEFAULT](https://github.com/edx/configuration/blob/master/playbooks/roles/mfe/defaults/main.yml#L95) variable, but the hard part is to figure out the `env_extra` variable for each MFE.

morenol commented:
@pomegranited thank you for your help to move this forward, I added the two last changes that you suggested and squashed everything in a single commit!

edx-webhook commented:
@morenol 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/configuration&entry.867055334=https://github.com/edx/configuration/pull/6171&entry.1484655318=https://github.com/morenol&entry.752974735=2020-11-24+22:57&entry.1917517419=2021-01-06+04:05&entry.2133058324=Yes) so we can improve your experience in the future.
